trin-core/src/utp/stream.rs

```diff
@@ -766,35 +844,337 @@ impl UtpSocket {
+    use std::sync::Arc;
+
+    fn next_test_port() -> u16 {
+        use std::sync::atomic::{AtomicUsize, Ordering};
+        static NEXT_OFFSET: AtomicUsize = AtomicUsize::new(0);
+        const BASE_PORT: u16 = 9600;
+        BASE_PORT + NEXT_OFFSET.fetch_add(1, Ordering::Relaxed) as u16
+    }
 
-            self.send_packet(&mut packet_reply);
+    fn create_portal_config() -> PortalnetConfig {
+        PortalnetConfig {
+            listen_port: next_test_port(),
+            internal_ip: true,
+            ..Default::default()
         }
     }
 
-    fn handle_syn_packet(&mut self, packet: Packet) {
-        self.receiver_connection_id = packet.connection_id() + 1;
-        self.sender_connection_id = packet.connection_id();
-        self.seq_nr = rand();
-        self.ack_nr = packet.seq_nr();
-        self.state = SocketState::SynRecv;
+    fn generate_random_enr() -> Enr {
+        let key = CombinedKey::generate_secp256k1();
+
+        let mut rng = rand::thread_rng();
+        let ip = Ipv4Addr::from(rng.gen::<u32>());
+
+        EnrBuilder::new("v4")
+            .ip(ip.into())
+            .udp(8000)
+            .build(&key)
+            .unwrap()
+    }
+
+    #[tokio::test]
+    async fn test_handle_packet() {
+        // Boilerplate test setup
+        let initial_connection_id: u16 = rand::random();
+        let sender_connection_id = initial_connection_id + 1;
+        let client_enr = generate_random_enr();
+        let server_enr = generate_random_enr();
+        let server_config = create_portal_config();
+        let mut server_discv5 = Discovery::new(server_config).unwrap();
+        server_discv5.start().await.unwrap();
+        let mut socket = UtpSocket::new(Arc::new(server_discv5), server_enr.clone());
+
+        // ---------------------------------
+        // Test connection setup - SYN packet
+
+        let mut packet = Packet::new();
+        packet.set_wnd_size(BUF_SIZE as u32);
+        packet.set_type(PacketType::Syn);
+        packet.set_connection_id(initial_connection_id);
+
+        // Do we have a response?
+        let response = socket.handle_packet(&packet, client_enr.clone());
+        assert!(response.is_ok());
+        let response = response.unwrap();
+        assert!(response.is_some());
 
-        let mut packet_reply = Packet::new();
-        packet_reply.set_type(PacketType::State);
-        packet_reply.set_connection_id(self.sender_connection_id);
-        packet_reply.set_seq_nr(self.seq_nr);
-        packet_reply.set_ack_nr(self.ack_nr);
+        // Is it of the correct type?
+        let response = response.unwrap();
+        assert_eq!(response.get_type(), PacketType::State);
 
-        self.send_packet(&mut packet_reply);
+        // Same connection id on both ends during connection establishment
+        assert_eq!(response.connection_id(), packet.connection_id());
+
+        // Response acknowledges SYN
+        assert_eq!(response.ack_nr(), packet.seq_nr());
+
+        // No payload?
```

Comment:
Is this a question? Is there supposed to be a payload? Something like `// Expect no payload` might be better